pkg/controller/sparkapplication/monitoring_config.go

```diff
@@ -39,50 +40,67 @@ const (
 )
 
 func configPrometheusMonitoring(app *v1beta1.SparkApplication, kubeClient clientset.Interface) error {
-	prometheusConfigMapName := fmt.Sprintf("%s-%s", app.Name, prometheusConfigMapNameSuffix)
-	configMap := buildPrometheusConfigMap(app, prometheusConfigMapName)
-	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		cm, err := kubeClient.CoreV1().ConfigMaps(app.Namespace).Get(prometheusConfigMapName, metav1.GetOptions{})
-
-		if apiErrors.IsNotFound(err) {
-			_, createErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Create(configMap)
-			return createErr
-		}
-		if err != nil {
-			return err
-		}
+	port := config.DefaultPrometheusJavaAgentPort
+	if app.Spec.Monitoring.Prometheus.Port != nil {
+		port = *app.Spec.Monitoring.Prometheus.Port
+	}
 
-		cm.Data = configMap.Data
-		_, updateErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Update(cm)
-		return updateErr
-	})
+	configFile := app.Spec.Monitoring.Prometheus.ConfigFile
+	var javaOption string
+	if configFile != "" {
+		glog.V(2).Infof("Overriding the default Prometheus configuration with config file %s in the Spark job image.", configFile)
+		javaOption = fmt.Sprintf("-javaagent:%s=%d:%s", app.Spec.Monitoring.Prometheus.JmxExporterJar,
+			port, configFile)
+	} else {
+		glog.V(2).Infof("Using the default Prometheus configuration.")
+		prometheusConfigMapName := fmt.Sprintf("%s-%s", app.Name, prometheusConfigMapNameSuffix)
+		configMap := buildPrometheusConfigMap(app, prometheusConfigMapName)
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			cm, err := kubeClient.CoreV1().ConfigMaps(app.Namespace).Get(prometheusConfigMapName, metav1.GetOptions{})
+
+			if apiErrors.IsNotFound(err) {
+				_, createErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Create(configMap)
+				return createErr
+			}
+			if err != nil {
+				return err
+			}
+
+			cm.Data = configMap.Data
+			_, updateErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Update(cm)
+			return updateErr
+		})
 
-	if retryErr != nil {
-		return fmt.Errorf("failed to apply %s in namespace %s: %v", prometheusConfigMapName, app.Namespace, retryErr)
+		if retryErr != nil {
+			return fmt.Errorf("failed to apply %s in namespace %s: %v", prometheusConfigMapName, app.Namespace, retryErr)
+		}
+		javaOption = fmt.Sprintf("-javaagent:%s=%d:%s/%s", app.Spec.Monitoring.Prometheus.JmxExporterJar,
```

Comment:
Please add an empty line before this line.

Reply:
Line 77 is restored.